### PR TITLE
Build optimizations for e2e and toolchain

### LINKED ISSUE #35517: Step building the same over and over again
While running make quick-release I get the following:

```
wojtekt@wojtekt-work:~/go/src/k8s.io/kubernetes$ make quick-release
+++ [1025 15:02:11] Verifying Prerequisites....
+++ [1025 15:02:12] Syncing sources to container
+++ [1025 15:02:13] Running build command...
+++ [1025 15:02:21] Generating bindata:
    /go/src/k8s.io/kubernetes/test/e2e/framework/gobindata_util.go
+++ [1025 15:02:22] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [1025 15:02:22] Building go targets for linux/amd64:
    cmd/libs/go2idl/deepcopy-gen
+++ [1025 15:02:26] Generating bindata:
    /go/src/k8s.io/kubernetes/test/e2e/framework/gobindata_util.go
+++ [1025 15:02:27] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [1025 15:02:27] Building go targets for linux/amd64:
    cmd/libs/go2idl/defaulter-gen
+++ [1025 15:02:31] Generating bindata:
    /go/src/k8s.io/kubernetes/test/e2e/framework/gobindata_util.go
+++ [1025 15:02:32] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [1025 15:02:32] Building go targets for linux/amd64:
    cmd/libs/go2idl/conversion-gen
+++ [1025 15:02:37] Generating bindata:
    /go/src/k8s.io/kubernetes/test/e2e/framework/gobindata_util.go
+++ [1025 15:02:38] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [1025 15:02:38] Building go targets for linux/amd64:
    cmd/kube-dns
    cmd/kube-proxy
    cmd/kube-apiserver
    cmd/kube-controller-manager
    cmd/kubelet
    cmd/kubeadm
    cmd/hyperkube
    cmd/kube-discovery
    plugin/cmd/kube-scheduler
+++ [1025 15:04:19] Generating bindata:
    /go/src/k8s.io/kubernetes/test/e2e/framework/gobindata_util.go
+++ [1025 15:04:20] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [1025 15:04:20] Building go targets for linux/amd64:
    cmd/libs/go2idl/deepcopy-gen
+++ [1025 15:04:26] Generating bindata:
    /go/src/k8s.io/kubernetes/test/e2e/framework/gobindata_util.go
+++ [1025 15:04:27] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [1025 15:04:27] Building go targets for linux/amd64:
    cmd/libs/go2idl/defaulter-gen
+++ [1025 15:04:31] Generating bindata:
    /go/src/k8s.io/kubernetes/test/e2e/framework/gobindata_util.go
+++ [1025 15:04:32] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [1025 15:04:32] Building go targets for linux/amd64:
    cmd/libs/go2idl/conversion-gen
+++ [1025 15:04:38] Generating bindata:
    /go/src/k8s.io/kubernetes/test/e2e/framework/gobindata_util.go
+++ [1025 15:04:39] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [1025 15:04:39] Building go targets for linux/amd64:
    cmd/kubectl
+++ [1025 15:05:14] Generating bindata:
    /go/src/k8s.io/kubernetes/test/e2e/framework/gobindata_util.go
+++ [1025 15:05:15] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [1025 15:05:15] Building go targets for linux/amd64:
    cmd/libs/go2idl/deepcopy-gen
+++ [1025 15:05:20] Generating bindata:
    /go/src/k8s.io/kubernetes/test/e2e/framework/gobindata_util.go
+++ [1025 15:05:21] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [1025 15:05:21] Building go targets for linux/amd64:
    cmd/libs/go2idl/defaulter-gen
+++ [1025 15:05:25] Generating bindata:
    /go/src/k8s.io/kubernetes/test/e2e/framework/gobindata_util.go
+++ [1025 15:05:26] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [1025 15:05:26] Building go targets for linux/amd64:
    cmd/libs/go2idl/conversion-gen
+++ [1025 15:05:31] Generating bindata:
    /go/src/k8s.io/kubernetes/test/e2e/framework/gobindata_util.go
+++ [1025 15:05:32] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [1025 15:05:32] Building go targets for linux/amd64:
    cmd/gendocs
    cmd/genkubedocs
    cmd/genman
    cmd/genyaml
    cmd/mungedocs
    cmd/genswaggertypedocs
    cmd/linkcheck
    examples/k8petstore/web-server/src
    federation/cmd/genfeddocs
    vendor/github.com/onsi/ginkgo/ginkgo
    test/e2e/e2e.test
+++ [1025 15:06:57] Generating bindata:
    /go/src/k8s.io/kubernetes/test/e2e/framework/gobindata_util.go
+++ [1025 15:06:58] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [1025 15:06:58] Building go targets for linux/amd64:
    cmd/libs/go2idl/deepcopy-gen
+++ [1025 15:07:03] Generating bindata:
    /go/src/k8s.io/kubernetes/test/e2e/framework/gobindata_util.go
+++ [1025 15:07:04] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [1025 15:07:04] Building go targets for linux/amd64:
    cmd/libs/go2idl/defaulter-gen
+++ [1025 15:07:08] Generating bindata:
    /go/src/k8s.io/kubernetes/test/e2e/framework/gobindata_util.go
+++ [1025 15:07:09] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [1025 15:07:09] Building go targets for linux/amd64:
    cmd/libs/go2idl/conversion-gen
+++ [1025 15:07:14] Generating bindata:
    /go/src/k8s.io/kubernetes/test/e2e/framework/gobindata_util.go
+++ [1025 15:07:14] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [1025 15:07:14] Building go targets for linux/amd64:
    cmd/kubemark
    vendor/github.com/onsi/ginkgo/ginkgo
...
```

We are building gobindata and teststale 16 times each, which clearly seems like a waste of time.

@thockin @gmarek @jayunit100

### PR DESCRIPTION
**Pull request by jayunit100**
Fixes #35517

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35553)

### CODE REVIEW

**Comment by jayunit100** [comment]
I think this should work... will see how build likes it.

**Comment by jayunit100** [comment]
will circle back to this later... definetely not simple enough to just grep e2e/cmd out.  will see why later.

**Comment by thockin** [comment]
@ixdy because overlap

**Comment by jayunit100** [comment]
@k8s-bot unit test this please

**Comment by jayunit100** [comment]
Ahhh i see, test-integration wants bindata (for some reason).  Digging into that... 

```
W1118 07:15:01.748] k8s.io/kubernetes/test/e2e/framework
W1118 07:15:03.650] # k8s.io/kubernetes/test/e2e/framework
W1118 07:15:03.651] test/e2e/framework/gobindata_util.go:30: undefined: generated.Asset
W1118 07:15:03.651] test/e2e/framework/gobindata_util.go:33: undefined: generated.AssetNames
W1118 07:15:03.735] make[1]: *** [test] Error 1
I1118 07:15:03.735] Makefile:126: recipe for target 'test' failed
I1118 07:15:03.739] +++ [1118 07:15:03] Cleaning up etcd
I1118 07:15:03.748] +++ [1118 07:15:03] Integration test cleanup complete
I1118 07:15:03.748] Makefile:138: recipe for target 'test-integration' failed
W1118 07:15:03.749] make: *** [test-integration] Error 1
I1118 07:15:04.724] process 1311 exited with code 2
E1118 07:15:04.725] FAIL: pull-kubernetes-unit
```

**Comment by jayunit100** [comment]
Updated manual generate step for `test-integration`

File: hack/lib/golang.sh [hunk 1]
```diff
@@ -610,16 +610,12 @@ kube::golang::build_binaries() {
...
-          # Only try to generate bindata if the file exists, since in some cases
-          # one-off builds of individual directories may exclude some files.
-      if [[ -f $bindata ]]; then
-          go generate "${bindata}"
-      fi
-    done
+    # TODO generalize this optimization somehow: Build bindata stuff is only really required for e2e as of now.
+    if [[ $@ == *"e2e"* ]]; then
+	 pushd `dirname test/e2e/`
+	 go generate ./...
```
Comment (ixdy) [inline hack/lib/golang.sh]: can we maybe instead make this a rule in `Makefile.generated_files`? may as well let `make` do some of the work for us.

**Review by ixdy — COMMENTED** [review]
yeah, we definitely need to see how this interacts with #36964.

File: hack/lib/golang.sh (hunk 1 above)
Comment (jayunit100) [inline hack/lib/golang.sh]: sure ... i can try that.  right now its mad at me about build files not being up to date, not sure why though

**Comment by jayunit100** [comment]
im going to close this.  fixing and stabilizing and cleaning the build is  a higher priority right now.

**Pull request closed without merging**